File: scala/src/test/scala/gameOfLife/GridManagerSpec.scala
package gameOfLife

import org.scalatest._

class GridManagerSpec extends FlatSpec with Matchers {
  "The GridManager" should "not change cells if all cells are dead" in {
    var gridManager: GridManager = new GridManager();
    var grid = Array.ofDim[Cell](1, 1);
    var newGrid = gridManager.GetNewState(grid);

    newGrid(0)(0).isAlive shouldEqual false;
  }

  it should "not kill cell if it has two neighbours" in {
    var gridManager = new GridManager();

    var grid = InitializeGrid(3, 3);
    grid(0)(2).isAlive = true;
    grid(1)(1).isAlive = true;
    grid(2)(0).isAlive = true;

    var newGrid = gridManager.GetNewState(grid);

    newGrid(1)(1).isAlive shouldEqual true;
  }

  it should "kill cell if it has less than two neighbours" in {
    var gridManager = new GridManager();

    var grid = InitializeGrid(2, 4);
    grid(1)(0).isAlive = true;
    grid(1)(1).isAlive = true;
    grid(1)(2).isAlive = true;

    var newGrid = gridManager.GetNewState(grid);

    newGrid(1)(0).isAlive shouldEqual false;
    newGrid(1)(2).isAlive shouldEqual false;
  }

  it should "alive cell if it has three neighbours" in {
    var gridManager = new GridManager();

    var grid = InitializeGrid(2, 3);
    grid(0)(1).isAlive = true;
    grid(1)(0).isAlive = true;
    grid(1)(1).isAlive = true;

    var newGrid = gridManager.GetNewState(grid);

    newGrid(0)(0).isAlive shouldEqual true;
  }

  it should "should kill cell if it has more than three neighbours" in {
    var gridManager = new GridManager();

    var grid = InitializeGrid(4, 3);
    grid(0)(2).isAlive = true;
    grid(1)(1).isAlive = true;
    grid(2)(0).isAlive = true;
    grid(2)(1).isAlive = true;
    grid(2)(2).isAlive = true;

    var newGrid = gridManager.GetNewState(grid);

    newGrid(1)(1).isAlive shouldEqual false;
  }

  it should "should count neighbours of cell" in {
    var gridManager = new GridManager();

    val h: Int = 2;
    val w: Int = 3;
    var grid = InitializeGrid(h, w);
    grid(0)(1).isAlive = true;
    grid(1)(0).isAlive = true;
    grid(1)(1).isAlive = true;

    var nbNeighbours = gridManager.CountNeighbours(grid, 0, 1);

    nbNeighbours shouldEqual 2;
  }

}

object InitializeGrid {
  def apply(h: Int, w: Int): Array[Array[Cell]] = {
    var grid = Array.ofDim[Cell](h, w);
    for (i <- 0 until grid.length) {
      for (j <- 0 until grid(0).length) {
        grid(i)(j) = new Cell(false);
      }
    }

    return grid;
  }
}

File: typescript/test/gameoflife/grid-manager.spec.ts
import GridManager from "../../src/gameoflife/grid-manager";
import Cell from "../../src/gameoflife/cell";

describe("GridManager", () => {

    it("should_not_change_cells_if_all_cells_are_dead", () => {
        const gridManager: GridManager = new GridManager();
        const grid: Cell[][] = [[new Cell(false)]];
        const newGrid: Cell[][] = gridManager.getNewState(grid);

        expect(newGrid[0][0].alive).toEqual(false);
    });


    it("should_not_kill_cell_if_it_has_two_neighbours", () => {
        const gridManager: GridManager = new GridManager();

        const grid = initializeGrid(3, 3);
        grid[0][2].alive = true;
        grid[1][1].alive = true;
        grid[2][0].alive = true;

        const newGrid: Cell[][] = gridManager.getNewState(grid);

        expect(newGrid[1][1].alive).toBe(true);
    });


    it("should_kill_cell_if_it_has_less_than_two_neighbours", () => {
        const gridManager: GridManager = new GridManager();

        const grid: any = initializeGrid(2, 4);
        grid[1][0].alive = true;
        grid[1][1].alive = true;
        grid[1][2].alive = true;

        const newGrid = gridManager.getNewState(grid);

        expect(newGrid[1][0].alive).toBe(false);
        expect(newGrid[1][2].alive).toBe(false);
    });


    it("should_alive_cell_if_it_has_three_neighbours", () => {
        const gridManager: GridManager = new GridManager();

        const grid: any = initializeGrid(2, 3);
        grid[0][1].alive = true;
        grid[1][0].alive = true;
        grid[1][1].alive = true;

        const newGrid: Cell[][] = gridManager.getNewState(grid);

        expect(newGrid[0][0].alive).toBe(true);
    });


    it("should_kill_cell_if_it_has_more_than_three_neighbours", () => {
        const gridManager: GridManager = new GridManager();

        const grid: Cell[][] = initializeGrid(4, 3);
        grid[0][2].alive = true;
        grid[1][1].alive = true;
        grid[2][0].alive = true;
        grid[2][1].alive = true;
        grid[2][2].alive = true;

        const newGrid: Cell[][] = gridManager.getNewState(grid);

        expect(newGrid[1][1].alive).toBe(false);
    });


    it("should_count_neighbours_of_cell", () => {
        const gridManager: GridManager = new GridManager();

        const h = 2;
        const w = 3;
        const grid = initializeGrid(h, w);
        grid[0][1].alive = true;
        grid[1][0].alive = true;
        grid[1][1].alive = true;

        const nbNeighbours = gridManager.countNeighbours(grid, 0, 1);

        expect(nbNeighbours).toEqual(2);
    });


});

function initializeGrid(h: number, w: number): Cell[][] {
    const grid: Cell[][] = new Array<Cell[]>(h);
    for (let i = 0; i < grid.length; i++) {
        grid[i] = new Array<Cell>(w);
        for (let j = 0; j < grid[0].length; j++) {
            grid[i][j] = new Cell(false);
        }
    }

    return grid;
}
